app/cdap/components/CreateAndEditRecipeForm/__test__/CreateAndEditRecipeForm.test.tsx

```diff
@@ -0,0 +1,95 @@
+ *  the License.
+ */
+
+import { render, screen, fireEvent } from '@testing-library/react';
+import React from 'react';
+import T from 'i18n-react';
+import CreateAndEditRecipeForm from 'components/CreateAndEditRecipeForm/index';
+import MyDataPrepApi from 'api/dataprep';
+
+describe('Test Create and Edit Recipe Component', () => {
+  const mockCreateAndEditRecipe = jest.fn();
+
+  jest.spyOn(MyDataPrepApi, 'updateRecipe').mockImplementation(() => {
+    return {
+      subscribe: (callback) => {
+        callback();
+      },
+    };
+  });
+
+  beforeEach(() => {
+    render(
+      <CreateAndEditRecipeForm
+        recipeData={{
+          recipeName: 'Abhilash',
+          description: 'Batman',
+          directives: [],
+          recipeId: { recipeId: '420', namespace: { name: 'test', generation: 69 } },
+        }}
+        setIsCreateAndEditRecipeFormOpen={mockCreateAndEditRecipe}
+        recipeFormAction={'editRecipe'}
+        setSnackbar={jest.fn()}
+      />
+    );
+  });
+
+  it('should render create and edit recipe form component', () => {
+    const parentElement = screen.getByTestId(/recipe-form-parent/i);
+    expect(parentElement).toBeInTheDocument();
+  });
+
+  it('should render Recipe Name Label', () => {
+    const RecipeNameLabelElement = screen.getByTestId(/recipe-name-label/i);
+    expect(RecipeNameLabelElement).toBeInTheDocument();
+    expect(RecipeNameLabelElement).toHaveTextContent(
+      `${T.translate('features.WranglerNewUI.RecipeForm.labels.name')}`
+    );
+  });
+
+  it('should render Recipe Description Label', () => {
+    const RecipeDescriptionLabelElement = screen.getByTestId(/recipe-description-label/i);
+    expect(RecipeDescriptionLabelElement).toBeInTheDocument();
+    expect(RecipeDescriptionLabelElement).toHaveTextContent(
+      `${T.translate('features.WranglerNewUI.RecipeForm.labels.description')}`
+    );
+  });
+
+  it('should render Recipe Name Field and trigger the change event as expected', () => {
+    const RecipeNameElement = screen.getByTestId(/recipe-name-field/i);
+    fireEvent.change(RecipeNameElement.firstChild.firstChild, { target: { value: 'test' } });
+    expect(RecipeNameElement).toBeInTheDocument();
+  });
+
+  it('should render Recipe Description Field', () => {
+    const RecipeDescriptionElement = screen.getByTestId(/recipe-description-field/i);
+    fireEvent.change(RecipeDescriptionElement, { target: { value: 'test' } });
+    expect(RecipeDescriptionElement).toHaveValue('test');
+  });
+
+  it('should trigger onCancel event in recipe', () => {
+    const cancelButtonElement = screen.getByTestId(/recipe-cancel-button/i);
+    fireEvent.click(cancelButtonElement);
+    expect(mockCreateAndEditRecipe).toBeCalled();
+  });
+
+  it('should trigger onSave event in recipe', () => {
+    const saveButtonElement = screen.getByTestId(/recipe-save-button/i);
+    fireEvent.click(saveButtonElement);
```

Comment:
Should these 2 statements be swapped i.e. checking for savebuttonelement before firing an event? Or is the intent of testing to ensure that savebuttonelement exists even after it was fired. Also, in case you swap them, do you foresee any change in the state of the button and should that be asserted here?

Reply:
The same applies for all other tests ?

Reply:
screen.getByTestId checks for the existence of the button by default. we can have the check before and after as well, to make sure the element is present before and after the interaction. Yes , the same applies for all tests.

Reply:
Added another assertion to check if the function related to that event is being called or not. Done.